File: client/Sources/Features/Agent/AgentSessionModel.swift
// client/Sources/Features/Agent/AgentSessionModel.swift
import Foundation
import MaverickProtocol

// MARK: - Chat timeline item

struct AgentChatItem: Identifiable {
    enum Kind {
        case userBubble(text: String)
        case assistantBubble(text: String, isStreaming: Bool)
        case toolBatch(tools: [ToolCallEvent], isCollapsed: Bool)
        case permissionRequest(event: PermissionEvent)
        case statusBadge(text: String, kind: BadgeKind)
        case turnSummary(cost: Double?, inputTokens: Int?, outputTokens: Int?, effortLevel: EffortLevel?)
        case sessionError(reason: StopFailureReason)
    }

    let id: UUID
    var kind: Kind

    init(id: UUID = UUID(), kind: Kind) {
        self.id = id
        self.kind = kind
    }
}

// MARK: - Per-session accumulator

@Observable
final class AgentSessionModel {
    let sessionId: UUID
    var provider: AgentProvider
    var mode: SessionMode
    var cwd: String
    var model: String?

    var isThinking: Bool = false
    var items: [AgentChatItem] = []

    /// Active permission request awaiting user input. Nil when none is pending.
    var pendingPermission: PermissionEvent? = nil

    private var pendingToolCalls: [String: ToolCallEvent] = [:]
    private var completedBatch: [ToolCallEvent] = []
    private var streamingBubbleId: UUID? = nil

    init(sessionId: UUID, provider: AgentProvider, mode: SessionMode, cwd: String) {
        self.sessionId = sessionId
        self.provider = provider
        self.mode = mode
        self.cwd = cwd
    }

    // MARK: - Public API

    /// Called by the UI after the user approves or denies a permission.
    func resolvePermission(requestId: String) {
        if pendingPermission?.requestId == requestId {
            pendingPermission = nil
        }
    }

    func apply(_ event: AgentEvent) {
        switch event {
        case .sessionStart(_, let provider, let cwd, let model, _):
            self.provider = provider
            self.cwd = cwd
            self.model = model

        case .cwdChanged(_, let to):
            cwd = to

        case .userMessage(let text):
            flushBatch()
            items.append(AgentChatItem(kind: .userBubble(text: text)))

        case .tokenDelta(let text):
            flushBatch()
            isThinking = false
            // Optimistic O(1) check: streaming bubble is always the last item.
            if let sid = streamingBubbleId,
               let last = items.last, last.id == sid,
               case .assistantBubble(let existing, _) = last.kind {
                items[items.count - 1].kind = .assistantBubble(text: existing + text, isStreaming: true)
            } else {
                let newId = UUID()
                streamingBubbleId = newId
                items.append(AgentChatItem(id: newId, kind: .assistantBubble(text: text, isStreaming: true)))
            }

        case .assistantMessage(let text):
            flushBatch()
            finalizeStreamingBubble()
            items.append(AgentChatItem(kind: .assistantBubble(text: text, isStreaming: false)))

        case .toolCallStart(let evt):
            isThinking = true
            pendingToolCalls[evt.id] = evt

        case .toolCallComplete(let evt):
            pendingToolCalls.removeValue(forKey: evt.id)
            completedBatch.append(evt)
            if pendingToolCalls.isEmpty { isThinking = false }

        case .toolCallFailed(let evt):
            pendingToolCalls.removeValue(forKey: evt.id)
            completedBatch.append(evt)
            if pendingToolCalls.isEmpty { isThinking = false }

        case .toolBatchComplete(let events):
            // Server-authoritative batch supersedes our locally accumulated list.
            completedBatch = []
            pendingToolCalls = [:]
            items.append(AgentChatItem(kind: .toolBatch(tools: events, isCollapsed: true)))
            isThinking = false

        case .permissionRequest(let evt):
            pendingPermission = evt
            items.append(AgentChatItem(kind: .permissionRequest(event: evt)))
            isThinking = false

        case .permissionDenied:
            pendingPermission = nil

        case .statusBadge(let text, let kind):
            items.append(AgentChatItem(kind: .statusBadge(text: text, kind: kind)))

        case .turnStop(let cost, let inputTokens, let outputTokens, let effortLevel):
            flushBatch()
            finalizeStreamingBubble()
            isThinking = false
            if cost != nil || inputTokens != nil || outputTokens != nil {
                items.append(AgentChatItem(kind: .turnSummary(
                    cost: cost, inputTokens: inputTokens,
                    outputTokens: outputTokens, effortLevel: effortLevel
                )))
            }

        case .sessionError(let reason):
            flushBatch()
            finalizeStreamingBubble()
            isThinking = false
            items.append(AgentChatItem(kind: .sessionError(reason: reason)))

        case .sessionEnd:
            flushBatch()
            finalizeStreamingBubble()
            isThinking = false

        default:
            break
        }
    }

    // MARK: - Private

    private func flushBatch() {
        guard !completedBatch.isEmpty else { return }
        let batch = completedBatch
        completedBatch = []
        items.append(AgentChatItem(kind: .toolBatch(tools: batch, isCollapsed: true)))
    }

    private func finalizeStreamingBubble() {
        guard let sid = streamingBubbleId else { return }
        streamingBubbleId = nil
        guard let last = items.last, last.id == sid,
              case .assistantBubble(let text, _) = last.kind else { return }
        items[items.count - 1].kind = .assistantBubble(text: text, isStreaming: false)
    }
}

File: mobile/src/stores/agent-session-model.ts
/**
 * Port of the Swift `AgentSessionModel` + `AgentChatItem`. This is the killer
 * feature's core: it folds the raw `AgentEvent` stream into a render-ready
 * timeline of chat items, preserving every invariant from Swift:
 *
 *   - token_delta appends to the trailing streaming assistant bubble (O(1)
 *     fast path; creates a new bubble if none is active)
 *   - assistant_message finalises any streaming bubble then appends a final one
 *   - tool_call_(start|complete|failed) accumulate into a pending batch; the
 *     batch flushes before the next user/assistant/turn boundary
 *   - tool_batch_complete is server-authoritative and supersedes the local list
 *   - permission_request sets pendingPermission AND appends an inline row
 *   - turn_stop appends a summary only when it carries cost/tokens
 *   - the streaming bubble is finalised before a flushed batch so a tool batch
 *     never interleaves inside an assistant segment
 *
 * History paging (loadAgentHistory): older items prepend at the head;
 * `trimResident` caps resident pages to ~5 (250 items) from the *head* so the
 * bottom-anchored live tail is never dropped.
 */

import {
  AgentEvent,
  AgentProvider,
  BadgeKind,
  EffortLevel,
  PermissionEvent,
  SessionMode,
  StopFailureReason,
  ToolCallEvent,
} from '@/protocol';
import { Observable } from '@/lib/observable';
import { randomUUID } from '@/protocol/primitives';

export type AgentChatItem =
  | { id: string; kind: 'user'; text: string }
  | { id: string; kind: 'assistant'; text: string; streaming: boolean }
  | { id: string; kind: 'toolBatch'; tools: ToolCallEvent[]; collapsed: boolean }
  | { id: string; kind: 'permission'; event: PermissionEvent }
  | { id: string; kind: 'statusBadge'; text: string; badge: BadgeKind }
  | {
      id: string;
      kind: 'turnSummary';
      cost?: number;
      inputTokens?: number;
      outputTokens?: number;
      effortLevel?: EffortLevel;
    }
  | { id: string; kind: 'sessionError'; reason: StopFailureReason };

export const HISTORY_PAGE_SIZE = 50;
export const MAX_RESIDENT_PAGES = 5;
const MAX_RESIDENT_ITEMS = HISTORY_PAGE_SIZE * MAX_RESIDENT_PAGES;

export class AgentSessionModel extends Observable {
  provider: AgentProvider;
  mode: SessionMode;
  cwd: string;
  model?: string;
  isThinking = false;

  private timeline: AgentChatItem[] = [];
  private pending: PermissionEvent | null = null;

  private readonly pendingToolCalls = new Map<string, ToolCallEvent>();
  private completedBatch: ToolCallEvent[] = [];
  private streamingBubbleId: string | null = null;

  constructor(
    readonly sessionId: string,
    provider: AgentProvider,
    mode: SessionMode,
    cwd: string,
  ) {
    super();
    this.provider = provider;
    this.mode = mode;
    this.cwd = cwd;
  }

  get items(): AgentChatItem[] {
    return this.timeline;
  }

  get pendingPermission(): PermissionEvent | null {
    return this.pending;
  }

  /** Called by the UI after the user approves or denies a permission. */
  resolvePermission(requestId: string): void {
    if (this.pending?.requestId === requestId) {
      this.pending = null;
      this.notify();
    }
  }

  /** Toggle a tool-batch row's collapsed state (UI affordance). */
  toggleBatch(itemId: string): void {
    const idx = this.timeline.findIndex((i) => i.id === itemId);
    if (idx < 0) return;
    const item = this.timeline[idx]!;
    if (item.kind !== 'toolBatch') return;
    this.timeline = this.timeline.slice();
    this.timeline[idx] = { ...item, collapsed: !item.collapsed };
    this.notify();
  }

  apply(event: AgentEvent): void {
    switch (event.type) {
      case 'session_start':
        this.provider = event.provider;
        this.cwd = event.cwd;
        this.model = event.model;
        break;
      case 'cwd_changed':
        this.cwd = event.to;
        break;
      case 'user_message':
        this.flushBatch();
        this.push({ id: randomUUID(), kind: 'user', text: event.text });
        break;
      case 'token_delta': {
        this.flushBatch();
        this.isThinking = false;
        const last = this.timeline[this.timeline.length - 1];
        if (
          this.streamingBubbleId !== null &&
          last !== undefined &&
          last.id === this.streamingBubbleId &&
          last.kind === 'assistant'
        ) {
          this.timeline = this.timeline.slice();
          this.timeline[this.timeline.length - 1] = {
            ...last,
            text: last.text + event.text,
            streaming: true,
          };
        } else {
          const id = randomUUID();
          this.streamingBubbleId = id;
          this.push({ id, kind: 'assistant', text: event.text, streaming: true });
        }
        break;
      }
      case 'assistant_message':
        this.flushBatch();
        this.finalizeStreamingBubble();
        this.push({
          id: randomUUID(),
          kind: 'assistant',
          text: event.text,
          streaming: false,
        });
        break;
      case 'tool_call_start':
        this.isThinking = true;
        this.pendingToolCalls.set(event.event.id, event.event);
        break;
      case 'tool_call_complete':
      case 'tool_call_failed':
        this.pendingToolCalls.delete(event.event.id);
        this.completedBatch.push(event.event);
        if (this.pendingToolCalls.size === 0) this.isThinking = false;
        this.notify();
        break;
      case 'tool_batch_complete':
        this.completedBatch = [];
        this.pendingToolCalls.clear();
        this.push({
          id: randomUUID(),
          kind: 'toolBatch',
          tools: event.events,
          collapsed: true,
        });
        this.isThinking = false;
        break;
      case 'permission_request':
        this.pending = event.permissionEvent;
        this.push({
          id: randomUUID(),
          kind: 'permission',
          event: event.permissionEvent,
        });
        this.isThinking = false;
        break;
      case 'permission_denied':
        this.pending = null;
        this.notify();
        break;
      case 'status_badge':
        this.push({
          id: randomUUID(),
          kind: 'statusBadge',
          text: event.text,
          badge: event.kind,
        });
        break;
      case 'turn_stop':
        this.flushBatch();
        this.finalizeStreamingBubble();
        this.isThinking = false;
        if (
          event.cost !== undefined ||
          event.inputTokens !== undefined ||
          event.outputTokens !== undefined
        ) {
          const item: AgentChatItem = {
            id: randomUUID(),
            kind: 'turnSummary',
          };
          if (event.cost !== undefined) item.cost = event.cost;
          if (event.inputTokens !== undefined) item.inputTokens = event.inputTokens;
          if (event.outputTokens !== undefined) {
            item.outputTokens = event.outputTokens;
          }
          if (event.effortLevel !== undefined) item.effortLevel = event.effortLevel;
          this.push(item);
        } else {
          this.notify();
        }
        break;
      case 'session_error':
        this.flushBatch();
        this.finalizeStreamingBubble();
        this.isThinking = false;
        this.push({
          id: randomUUID(),
          kind: 'sessionError',
          reason: event.reason,
        });
        break;
      case 'session_end':
        this.flushBatch();
        this.finalizeStreamingBubble();
        this.isThinking = false;
        this.notify();
        break;
      default:
        // subagent/task/compaction/worktree/notification/elicitation/raw bytes
        // do not contribute chat items in the mobile timeline.
        this.notify();
        break;
    }
  }

  /**
   * Prepend a page of older, already-reduced items at the head (bottom-anchored
   * paging). Returns true if anything was prepended. Trims from the head if the
   * resident window exceeds MAX_RESIDENT_ITEMS.
   */
  prependHistory(older: AgentChatItem[]): boolean {
    if (older.length === 0) return false;
    this.timeline = [...older, ...this.timeline];
    this.trimResident();
    this.notify();
    return true;
  }

  private trimResident(): void {
    if (this.timeline.length > MAX_RESIDENT_ITEMS) {
      this.timeline = this.timeline.slice(
        this.timeline.length - MAX_RESIDENT_ITEMS,
      );
    }
  }

  private push(item: AgentChatItem): void {
    this.timeline = [...this.timeline, item];
    this.trimResident();
    this.notify();
  }

  private flushBatch(): void {
    if (this.completedBatch.length === 0) return;
    if (this.streamingBubbleId !== null) this.finalizeStreamingBubble();
    const batch = this.completedBatch;
    this.completedBatch = [];
    this.timeline = [
      ...this.timeline,
      { id: randomUUID(), kind: 'toolBatch', tools: batch, collapsed: true },
    ];
    this.trimResident();
  }

  private finalizeStreamingBubble(): void {
    const sid = this.streamingBubbleId;
    if (sid === null) return;
    this.streamingBubbleId = null;
    const last = this.timeline[this.timeline.length - 1];
    if (last !== undefined && last.id === sid && last.kind === 'assistant') {
      this.timeline = this.timeline.slice();
      this.timeline[this.timeline.length - 1] = { ...last, streaming: false };
      return;
    }
    // Fallback: the streaming bubble was pushed off the tail (a server
    // tool_batch_complete appends directly without finalising stream first).
    const idx = this.timeline.findIndex((i) => i.id === sid);
    /* istanbul ignore else -- the id always points at the assistant bubble we
       created; the guard is defensive. */
    if (idx >= 0) {
      const item = this.timeline[idx]!;
      /* istanbul ignore else -- same invariant: the item at sid is assistant. */
      if (item.kind === 'assistant') {
        this.timeline = this.timeline.slice();
        this.timeline[idx] = { ...item, streaming: false };
      }
    }
  }
}
